File: ios/ar_charts/Sources/ar_charts/BarChartPlatformView.swift
import Flutter
import UIKit
import DGCharts

private let defaultAxisLabelColor = UIColor(argb: 0xFF333333)

final class BarChartViewFactory: NSObject, FlutterPlatformViewFactory {
    func create(
        withFrame frame: CGRect,
        viewIdentifier viewId: Int64,
        arguments args: Any?
    ) -> FlutterPlatformView {
        let params = args as? [String: Any] ?? [:]
        return BarChartPlatformView(frame: frame, viewId: viewId, params: params)
    }

    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

final class BarChartPlatformView: NSObject, FlutterPlatformView, ChartMarkerSupporting {
    private let chartView: BarChartView
    private let viewId: Int64
    var markerAutoHideSeconds: TimeInterval?
    var autoHideWorkItem: DispatchWorkItem?

    var chartViewForMarker: BarLineChartViewBase { chartView }

    init(frame: CGRect, viewId: Int64, params: [String: Any]) {
        self.chartView = BarChartView(frame: frame)
        self.viewId = viewId
        super.init()
        ChartViewRegistry.registerBarChart(viewId: viewId, view: self)
        applyConfig(params: params)
    }

    deinit {
        cancelMarkerAutoHide()
        ChartViewRegistry.unregisterBarChart(viewId: viewId)
    }

    func view() -> UIView {
        return chartView
    }

    func updateConfig(params: [String: Any]) {
        applyConfig(params: params)
    }

    private func applyConfig(params: [String: Any]) {
        let dataSetsPayload = params["dataSets"] as? [[String: Any]] ?? []
        let defaultStyle = params["defaultBarStyle"] as? [String: Any]
        let perSeriesStyle = params["perSeriesStyle"] as? [String: Any]
        let group = params["group"] as? [String: Any]
        let groupEnabled = group?["enabled"] as? Bool ?? false
        let groupedXAxisContext = groupEnabled ? buildGroupedXAxisContext(dataSetsPayload: dataSetsPayload) : nil

        var dataSets: [BarChartDataSet] = []
        var barWidth: Double?
        var labelMap: [Int: String] = [:]
        for (index, item) in dataSetsPayload.enumerated() {
            let points = item["points"] as? [[String: Any]] ?? []
            let entries: [BarChartDataEntry]
            if groupEnabled, let ctx = groupedXAxisContext {
                // groupBars positions by entry index: index i → group i. Emit entries in group order.
                var pointByX: [Double: Double] = [:]
                for point in points {
                    guard let x = point["x"] as? Double, let y = point["y"] as? Double else { continue }
                    pointByX[x] = y
                }
                entries = ctx.sourceValuesInOrder.enumerated().map { groupIndex, sourceX in
                    let y = pointByX[sourceX] ?? 0
                    let xLabel = ctx.groupLabels[safe: groupIndex] ?? formatXAxisValue(sourceX)
                    return BarChartDataEntry(
                        x: Double(groupIndex),
                        y: y,
                        data: ["sourceX": sourceX, "xLabel": xLabel] as NSDictionary
                    )
                }
            } else {
                entries = points.compactMap { point in
                    guard let x = point["x"] as? Double, let y = point["y"] as? Double else { return nil }
                    if index == 0, let label = point["label"] as? String {
                        labelMap[Int(x)] = label
                    }
                    return BarChartDataEntry(x: x, y: y)
                }
            }
            let label = item["label"] as? String ?? ""
            let dataSet = BarChartDataSet(entries: entries, label: label)

            let seriesId = item["id"] as? String
            let styleMap = seriesId.flatMap { perSeriesStyle?[$0] as? [String: Any] } ?? defaultStyle
            let styleWidth = applyBarStyle(dataSet: dataSet, styleMap: styleMap)
            if barWidth == nil, let styleWidth {
                barWidth = styleWidth
            }
            dataSets.append(dataSet)
        }

        let data = BarChartData(dataSets: dataSets)
        if let barWidth {
            data.barWidth = barWidth
        }
        chartView.data = data

        let xAxisMap = params["xAxis"] as? [String: Any]
        applyAxis(axis: chartView.xAxis, axisMap: xAxisMap, isXAxis: true)
        applyAxis(axis: chartView.leftAxis, axisMap: params["leftAxis"] as? [String: Any], isXAxis: false)
        applyAxis(axis: chartView.rightAxis, axisMap: params["rightAxis"] as? [String: Any], isXAxis: false)

        if groupEnabled {
            applyGroupedBars(barGroup: group, data: data, groupedXAxisContext: groupedXAxisContext)
        } else {
            applyBarPointLabels(axis: chartView.xAxis, labelMap: labelMap)
        }

        applyLegend(legendMap: params["legend"] as? [String: Any])
        applyInteraction(interactionMap: params["interaction"] as? [String: Any])
        applyMarker(markerMap: params["marker"] as? [String: Any])
        applyAnimation(animationMap: params["animation"] as? [String: Any])

        chartView.chartDescription.enabled = false
        chartView.notifyDataSetChanged()
    }

    private func applyBarStyle(dataSet: BarChartDataSet, styleMap: [String: Any]?) -> Double? {
        guard let styleMap else { return nil }

        if let barColor = styleMap["barColor"] as? Int {
            dataSet.colors = [UIColor(argb: barColor)]
        }
        if let drawValues = styleMap["drawValues"] as? Bool {
            dataSet.drawValuesEnabled = drawValues
        }
        return styleMap["barWidth"] as? Double
    }

    private func applyAxis(axis: AxisBase, axisMap: [String: Any]?, isXAxis: Bool) {
        guard let axisMap else { return }
        axis.enabled = axisMap["enabled"] as? Bool ?? true
        axis.labelTextColor = defaultAxisLabelColor
        axis.drawGridLinesEnabled = axisMap["drawGridLines"] as? Bool ?? true
        if let gridLineColor = axisMap["gridLineColor"] as? Int {
            axis.gridColor = UIColor(argb: gridLineColor)
        } else {
            axis.gridColor = axis.labelTextColor.withAlphaComponent(0.22)
        }
        if let gridLineWidth = axisMap["gridLineWidth"] as? Double {
            axis.gridLineWidth = gridLineWidth
        } else if let gridLineWidth = axisMap["gridLineWidth"] as? Int {
            axis.gridLineWidth = Double(gridLineWidth)
        }
        axis.drawAxisLineEnabled = axisMap["drawAxisLine"] as? Bool ?? true

        if let min = axisMap["min"] as? Double { axis.axisMinimum = min }
        if let max = axisMap["max"] as? Double { axis.axisMaximum = max }
        if let labelCount = axisMap["labelCount"] as? Int { axis.setLabelCount(labelCount, force: true) }
        if let xAxis = axis as? XAxis, isXAxis {
            xAxis.labelPosition = .bottom
        }
        if let formatter = YAxisValueFormatter(axisMap: axisMap) {
            axis.valueFormatter = formatter
        }
    }

    private func applyBarPointLabels(axis: XAxis, labelMap: [Int: String]) {
        guard let maxIndex = labelMap.keys.max(), !labelMap.isEmpty else { return }
        var labels = (0...maxIndex).map { String($0) }
        for (index, label) in labelMap {
            if index >= 0 && index < labels.count {
                labels[index] = label
            }
        }
        axis.valueFormatter = IndexAxisValueFormatter(values: labels)
        axis.granularity = 1
        axis.granularityEnabled = true
    }

    private func applyGroupedBars(
        barGroup: [String: Any]?,
        data: BarChartData,
        groupedXAxisContext: GroupedXAxisContext?
    ) {
        guard let barGroup, data.dataSetCount > 1 else { return }
        let groupSpace = barGroup["groupSpace"] as? Double ?? 0.2
        let barSpace = barGroup["barSpace"] as? Double ?? 0.05
        let fromX = barGroup["fromX"] as? Double ?? 0
        let centerAxisLabels = barGroup["centerAxisLabels"] as? Bool ?? true

        data.groupBars(fromX: fromX, groupSpace: groupSpace, barSpace: barSpace)
        chartView.xAxis.centerAxisLabelsEnabled = centerAxisLabels
        chartView.xAxis.valueFormatter = IndexAxisValueFormatter(values: groupedXAxisContext?.groupLabels ?? [])
        chartView.xAxis.granularity = 1
        chartView.xAxis.granularityEnabled = true

        let groupCount = groupedXAxisContext?.groupLabels.count ?? data.dataSets.first?.entryCount ?? 0
        let groupWidth = data.groupWidth(groupSpace: groupSpace, barSpace: barSpace)
        chartView.xAxis.axisMinimum = fromX
        chartView.xAxis.axisMaximum = fromX + groupWidth * Double(groupCount)
    }

    private func buildGroupedXAxisContext(dataSetsPayload: [[String: Any]]) -> GroupedXAxisContext {
        var pointsByX: [Double: String] = [:]
        for item in dataSetsPayload {
            let points = item["points"] as? [[String: Any]] ?? []
            for point in points {
                guard let x = point["x"] as? Double else { continue }
                if pointsByX[x] != nil { continue }
                pointsByX[x] = (point["label"] as? String) ?? formatXAxisValue(x)
            }
        }
        let sourceValuesInOrder = pointsByX.keys.sorted()
        let xToGroupIndex = Dictionary(uniqueKeysWithValues: sourceValuesInOrder.enumerated().map { ($0.element, $0.offset) })
        let groupLabels = sourceValuesInOrder.map { pointsByX[$0] ?? formatXAxisValue($0) }
        return GroupedXAxisContext(xToGroupIndex: xToGroupIndex, groupLabels: groupLabels, sourceValuesInOrder: sourceValuesInOrder)
    }

    private func formatXAxisValue(_ value: Double) -> String {
        if value == value.rounded() { return String(Int(value)) }
        return String(value)
    }

    private func applyLegend(legendMap: [String: Any]?) {
        guard let legendMap else { return }
        let legend = chartView.legend
        legend.enabled = legendMap["enabled"] as? Bool ?? true
        legend.textColor = defaultAxisLabelColor

        switch legendMap["position"] as? String {
        case "bottom":
            legend.verticalAlignment = .bottom
        case "left":
            legend.verticalAlignment = .top
            legend.horizontalAlignment = .left
        case "right":
            legend.verticalAlignment = .top
            legend.horizontalAlignment = .right
        default:
            legend.verticalAlignment = .top
            legend.horizontalAlignment = .center
        }

        switch legendMap["alignment"] as? String {
        case "start":
            legend.horizontalAlignment = .left
        case "end":
            legend.horizontalAlignment = .right
        default:
            legend.horizontalAlignment = .center
        }

        legend.orientation = .horizontal
        legend.drawInside = false
    }

    private func applyInteraction(interactionMap: [String: Any]?) {
        guard let interactionMap else { return }
        let zoomEnabled = interactionMap["zoomEnabled"] as? Bool ?? true
        let dragEnabled = interactionMap["dragEnabled"] as? Bool ?? true
        let highlightEnabled = interactionMap["highlightEnabled"] as? Bool ?? true

        chartView.scaleXEnabled = zoomEnabled
        chartView.scaleYEnabled = zoomEnabled
        chartView.doubleTapToZoomEnabled = zoomEnabled
        chartView.dragEnabled = dragEnabled
        chartView.highlightPerTapEnabled = highlightEnabled
    }

    private func applyAnimation(animationMap: [String: Any]?) {
        guard let animationMap else { return }
        let enabled = animationMap["enabled"] as? Bool ?? false
        if !enabled { return }
        let durationMs = animationMap["durationMs"] as? Double ?? 500
        let duration = durationMs / 1000.0
        switch animationMap["easing"] as? String {
        case "linear":
            chartView.animate(xAxisDuration: duration, easingOption: .linear)
        default:
            chartView.animate(xAxisDuration: duration, easingOption: .easeInOutQuad)
        }
    }
}

// MARK: - ChartMarkerSupporting (ChartViewDelegate)
extension BarChartPlatformView {
    func chartValueSelected(_ chartView: ChartViewBase, entry: ChartDataEntry, highlight: Highlight) {
        handleChartValueSelected(chartView, entry: entry, highlight: highlight)
    }

    func chartValueNothingSelected(_ chartView: ChartViewBase) {
        handleChartValueNothingSelected(chartView)
    }
}

private struct GroupedXAxisContext {
    let xToGroupIndex: [Double: Int]
    let groupLabels: [String]
    let sourceValuesInOrder: [Double]
}

private extension Array {
    subscript(safe index: Int) -> Element? {
        guard index >= 0, index < count else { return nil }
        return self[index]
    }
}

File: android/src/main/kotlin/com/anisrehman/archarts/BarChartPlatformView.kt
package com.anisrehman.archarts

import android.content.Context
import android.view.View
import com.github.mikephil.charting.charts.BarChart
import com.github.mikephil.charting.components.Legend
import com.github.mikephil.charting.components.XAxis
import com.github.mikephil.charting.components.YAxis
import com.github.mikephil.charting.data.BarData
import com.github.mikephil.charting.data.BarDataSet
import com.github.mikephil.charting.data.BarEntry
import com.github.mikephil.charting.formatter.IndexAxisValueFormatter
import com.github.mikephil.charting.highlight.Highlight
import io.flutter.plugin.platform.PlatformView

class BarChartPlatformView(
    context: Context,
    private val params: Map<String, Any?>
) : PlatformView {

    private val chart: BarChart = BarChart(context)

    init {
        applyConfig()
    }

    override fun getView(): View = chart

    override fun dispose() = Unit

    private fun applyConfig() {
        val series = params["series"] as? List<*> ?: emptyList<Any>()
        val defaultStyle = params["defaultBarStyle"] as? Map<String, Any?>
        val perSeriesStyle = params["perSeriesStyle"] as? Map<String, Any?>
        val barGroup = params["barGroup"] as? Map<String, Any?>
        val groupEnabled = barGroup?.get("enabled") as? Boolean ?: false

        var barWidth: Float? = null
        val labelMap = linkedMapOf<Int, String>()
        val dataSets = series.mapIndexedNotNull { index, item ->
            val seriesMap = item as? Map<String, Any?> ?: return@mapIndexedNotNull null
            val points = seriesMap["points"] as? List<*> ?: emptyList<Any>()
            val entries = points.mapNotNull { point ->
                val pointMap = point as? Map<String, Any?> ?: return@mapNotNull null
                val x = (pointMap["x"] as? Number)?.toFloat() ?: return@mapNotNull null
                val y = (pointMap["y"] as? Number)?.toFloat() ?: return@mapNotNull null
                if (!groupEnabled && index == 0) {
                    val label = pointMap["label"] as? String
                    if (label != null) {
                        labelMap[x.toInt()] = label
                    }
                }
                BarEntry(x, y)
            }
            val label = seriesMap["label"] as? String
            val dataSet = BarDataSet(entries, label ?: "")
            val seriesId = seriesMap["id"] as? String
            val styleMap = seriesId?.let { id -> perSeriesStyle?.get(id) as? Map<String, Any?> }
                ?: defaultStyle
            val styleWidth = applyBarStyle(dataSet, styleMap)
            if (barWidth == null && styleWidth != null) {
                barWidth = styleWidth
            }
            dataSet
        }

        val data = BarData(dataSets)
        if (barWidth != null) {
            data.barWidth = barWidth ?: data.barWidth
        }
        chart.data = data

        val xAxisMap = params["xAxis"] as? Map<String, Any?>
        applyAxis(chart.xAxis, xAxisMap, AxisType.X)
        applyAxis(chart.axisLeft, params["leftAxis"] as? Map<String, Any?>, AxisType.Y)
        applyAxis(chart.axisRight, params["rightAxis"] as? Map<String, Any?>, AxisType.Y)

        if (groupEnabled) {
            applyGroupedBars(chart.xAxis, barGroup, data, dataSets)
        } else {
            applyBarPointLabels(chart.xAxis, labelMap)
        }

        applyLegend(params["legend"] as? Map<String, Any?>)
        applyInteraction(params["interaction"] as? Map<String, Any?>)
        applyMarker(params["marker"] as? Map<String, Any?>)
        applyAnimation(params["animation"] as? Map<String, Any?>)

        chart.description.isEnabled = false
        chart.invalidate()
    }

    private fun applyBarStyle(dataSet: BarDataSet, styleMap: Map<String, Any?>?): Float? {
        if (styleMap == null) return null

        val barColor = (styleMap["barColor"] as? Number)?.toInt()
        if (barColor != null) {
            dataSet.color = barColor
        }
        val drawValues = styleMap["drawValues"] as? Boolean
        if (drawValues != null) {
            dataSet.setDrawValues(drawValues)
        }
        return (styleMap["barWidth"] as? Number)?.toFloat()
    }

    private fun applyAxis(axis: XAxis, axisMap: Map<String, Any?>?, type: AxisType) {
        if (axisMap == null) return
        axis.isEnabled = axisMap["enabled"] as? Boolean ?: true
        axis.setDrawGridLines(axisMap["drawGridLines"] as? Boolean ?: true)
        axis.setDrawAxisLine(axisMap["drawAxisLine"] as? Boolean ?: true)

        val min = (axisMap["min"] as? Number)?.toFloat()
        val max = (axisMap["max"] as? Number)?.toFloat()
        if (min != null) axis.axisMinimum = min
        if (max != null) axis.axisMaximum = max

        val labelCount = axisMap["labelCount"] as? Int
        if (labelCount != null) {
            axis.setLabelCount(labelCount, true)
        }
        if (type == AxisType.X) {
            axis.position = XAxis.XAxisPosition.BOTTOM
        }
        YAxisValueFormatter.fromAxisMap(axisMap)?.let { formatter ->
            axis.valueFormatter = formatter
        }
    }

    private fun applyAxis(axis: YAxis, axisMap: Map<String, Any?>?, type: AxisType) {
        if (axisMap == null) return
        axis.isEnabled = axisMap["enabled"] as? Boolean ?: true
        axis.setDrawGridLines(axisMap["drawGridLines"] as? Boolean ?: true)
        axis.setDrawAxisLine(axisMap["drawAxisLine"] as? Boolean ?: true)

        val min = (axisMap["min"] as? Number)?.toFloat()
        val max = (axisMap["max"] as? Number)?.toFloat()
        if (min != null) axis.axisMinimum = min
        if (max != null) axis.axisMaximum = max

        val labelCount = axisMap["labelCount"] as? Int
        if (labelCount != null) {
            axis.setLabelCount(labelCount, true)
        }

        YAxisValueFormatter.fromAxisMap(axisMap)?.let { formatter ->
            axis.valueFormatter = formatter
        }
    }

    private fun applyBarPointLabels(axis: XAxis, labelMap: Map<Int, String>) {
        if (labelMap.isEmpty()) return
        val maxIndex = labelMap.keys.maxOrNull() ?: return
        val labels = MutableList(maxIndex + 1) { index -> index.toString() }
        labelMap.forEach { (index, label) ->
            if (index in labels.indices) {
                labels[index] = label
            }
        }
        axis.valueFormatter = IndexAxisValueFormatter(labels)
        axis.granularity = 1f
    }

    private fun applyGroupedBars(
        axis: XAxis,
        barGroup: Map<String, Any?>?,
        data: BarData,
        dataSets: List<BarDataSet>
    ) {
        if (barGroup == null || dataSets.size < 2) return
        val groupSpace = (barGroup["groupSpace"] as? Number)?.toFloat() ?: 0.2f
        val barSpace = (barGroup["barSpace"] as? Number)?.toFloat() ?: 0.05f
        val fromX = (barGroup["fromX"] as? Number)?.toFloat() ?: 0f
        val centerAxisLabels = barGroup["centerAxisLabels"] as? Boolean ?: true

        data.groupBars(fromX, groupSpace, barSpace)
        axis.setCenterAxisLabels(centerAxisLabels)
        axis.granularity = 1f

        val groupCount = dataSets.firstOrNull()?.entryCount ?: 0
        val groupWidth = data.getGroupWidth(groupSpace, barSpace)
        axis.axisMinimum = fromX
        axis.axisMaximum = fromX + groupWidth * groupCount
    }

    private fun applyLegend(legendMap: Map<String, Any?>?) {
        val legend = chart.legend
        if (legendMap == null) return
        legend.isEnabled = legendMap["enabled"] as? Boolean ?: true

        when (legendMap["position"] as? String) {
            "bottom" -> legend.verticalAlignment = Legend.LegendVerticalAlignment.BOTTOM
            "left" -> legend.verticalAlignment = Legend.LegendVerticalAlignment.TOP
            "right" -> legend.verticalAlignment = Legend.LegendVerticalAlignment.TOP
            else -> legend.verticalAlignment = Legend.LegendVerticalAlignment.TOP
        }

        when (legendMap["position"] as? String) {
            "left" -> legend.horizontalAlignment = Legend.LegendHorizontalAlignment.LEFT
            "right" -> legend.horizontalAlignment = Legend.LegendHorizontalAlignment.RIGHT
            else -> legend.horizontalAlignment = Legend.LegendHorizontalAlignment.CENTER
        }

        legend.orientation = Legend.LegendOrientation.HORIZONTAL
        legend.setDrawInside(false)
    }

    private fun applyInteraction(interactionMap: Map<String, Any?>?) {
        if (interactionMap == null) return
        val zoomEnabled = interactionMap["zoomEnabled"] as? Boolean ?: true
        val dragEnabled = interactionMap["dragEnabled"] as? Boolean ?: true
        val highlightEnabled = interactionMap["highlightEnabled"] as? Boolean ?: true

        chart.setScaleEnabled(zoomEnabled)
        chart.setPinchZoom(zoomEnabled)
        chart.isDragEnabled = dragEnabled
        chart.isHighlightPerTapEnabled = highlightEnabled
        chart.isHighlightPerDragEnabled = highlightEnabled
    }

    private fun applyMarker(markerMap: Map<String, Any?>?) {
        if (markerMap == null) return
        val enabled = markerMap["enabled"] as? Boolean ?: false
        if (!enabled) return
        val format = markerMap["format"] as? String
        val marker = ChartMarkerView(chart.context, format)
        marker.chartView = chart
        chart.marker = marker
    }

    private fun applyAnimation(animationMap: Map<String, Any?>?) {
        if (animationMap == null) return
        val enabled = animationMap["enabled"] as? Boolean ?: false
        if (!enabled) return
        val duration = (animationMap["durationMs"] as? Number)?.toInt() ?: 500
        when (animationMap["easing"] as? String) {
            "linear" -> chart.animateX(duration, com.github.mikephil.charting.animation.Easing.Linear)
            else -> chart.animateX(duration, com.github.mikephil.charting.animation.Easing.EaseInOutQuad)
        }
    }

    private enum class AxisType { X, Y }
}
